tools/cmake/FindOpenBLAS.cmake

```diff
@@ -0,0 +1,61 @@
+SET(Open_BLAS_INCLUDE_SEARCH_PATHS
+  /usr/include
+  /usr/include/openblas
+  /usr/include/openblas-base
+  /usr/local/include
+  /usr/local/include/openblas
+  /usr/local/include/openblas-base
+  /opt/OpenBLAS/include
+  $ENV{OpenBLAS_HOME}
+  $ENV{OpenBLAS_HOME}/include
+  )
+
+SET(Open_BLAS_LIB_SEARCH_PATHS
+  $ENV{OpenBLAS}cd
+  $ENV{OpenBLAS}/lib
+  $ENV{OpenBLAS_HOME}
+  $ENV{OpenBLAS_HOME}/lib
+  /lib/
+  /lib/openblas-base
+  /lib64/
+  /usr/lib
+  /usr/lib/openblas-base
+  /usr/lib64
+  /usr/local/lib
+  /usr/local/lib64
+  /opt/OpenBLAS/lib
+  )
```

Comment:
Question: does CMake support searching for includes/libs the way autoconf does, i. e. compiling or linking one-liners to see if they succeed? Basically, this: https://github.com/kaldi-asr/kaldi/issues/3228. We can extend this list as infinitum (I ended up with the OpenBLAS package in the path like /opt/rh/epel/OpenBLAS on CentOS, IIR the path correctly), but the toolset is usually the best guy to let figure out it's own libraries. If linking with `-lopenblas` without an explicit `-L` path works, I do not want to even know where the library actually is; unless something is really borked in the user's OS configuration, I fully expect ld to find the .so library at runtime if linker found it without the -L (yes, the mechanisms are disconnected, but should better be consistent, or most program would break anyway). So I just do not want to take this work from the system; the less maintenance I do, the better. 

Also, regarding OpenBLAS only, I am hacking configure now in an assumption that the best OpenBLAS is in our tools/OpenBLAS/install/{include,lib} provided it exists, as it is the one the most optimized for the user's system. It also drops some files into the cmake/ subdirectory of one of these (I do not remember which, sorry, I'll check), apparently to help CMake builds find it.

Reply:
FYI, if you find it helpful, here's the contents of the two files in that `libs/cmake/` directory that is made by our `cd tools; make openblas`:
```
~/work/kaldi2/tools$ head -42 OpenBLAS/install/lib/cmake/openblas/*
==> OpenBLAS/install/lib/cmake/openblas/OpenBLASConfig.cmake <==
SET(OpenBLAS_VERSION "0.3.5")
SET(OpenBLAS_INCLUDE_DIRS /home/kkm/work/kaldi2/tools/OpenBLAS/install/include)
SET(OpenBLAS_LIBRARIES /home/kkm/work/kaldi2/tools/OpenBLAS/install/lib/libopenblas.so)

==> OpenBLAS/install/lib/cmake/openblas/OpenBLASConfigVersion.cmake <==
set (PACKAGE_VERSION "0.3.5")
if (PACKAGE_VERSION VERSION_LESS PACKAGE_FIND_VERSION)
  set (PACKAGE_VERSION_COMPATIBLE FALSE)
else ()
  set (PACKAGE_VERSION_COMPATIBLE TRUE)
  if (PACKAGE_FIND_VERSION STREQUAL PACKAGE_VERSION)
    set (PACKAGE_VERSION_EXACT TRUE)
  endif ()
endif ()
```
What I do not see here is a pointer to the .a version of the library. The build builds both .so and .a,  but these files which (I assume) help cmake discover the package do not mention the available static lib. We support `--static-math` and `--dynamic-math` that selects one of the two, and I think that keeping this support may be important (i. e., if you roll out kaldi executables onto a cluster and would rather be them self-contained, even if large). The libraries in the `/lib` for your reference:
```
~/work/kaldi2/tools$ ll OpenBLAS/install/lib/lib*
lrwxrwxrwx 1 kkm kkm  29 2019-04-14 22:27:38 OpenBLAS/install/lib/libopenblas.a -> libopenblas_skylakex-r0.3.5.a
-rw-r--r-- 1 kkm kkm 25M 2019-04-14 22:27:26 OpenBLAS/install/lib/libopenblas_skylakex-r0.3.5.a
-rwxr-xr-x 1 kkm kkm 13M 2019-04-14 22:27:31 OpenBLAS/install/lib/libopenblas_skylakex-r0.3.5.so*
lrwxrwxrwx 1 kkm kkm  30 2019-04-14 22:27:38 OpenBLAS/install/lib/libopenblas.so -> libopenblas_skylakex-r0.3.5.so*
lrwxrwxrwx 1 kkm kkm  30 2019-04-14 22:27:38 OpenBLAS/install/lib/libopenblas.so.0 -> libopenblas_skylakex-r0.3.5.so*
```

Maybe sending them a PR that would fix that (if this is not canonical, in cmake sense) would get it in in time you'd have a chance to continue this work, so at the least no `OpenBLAS_LIBRARIES` cutting-splitting-testing if .a exists would not be required.

These files are made by their standard Makefile build from the latest release 0.3.5, not the development branch.

Reply:
for a library named foo
1. if /usr/share/cmake-3.5/Modules/FindFOO.cmake, exists, or
2. if foo is a cmake project, or
3. if FooConfig.cmake is provided by author
then you should not write a FindFOO.cmake yourself